### PR TITLE
System.IO.Stream: Decorate validation helpers with [NotNull] to fix CA1062 false positives

### PR DESCRIPTION
**Pull request by CodeBlanch**
I have a custom stream implementation doing something like...

```csharp
public sealed class MyStream : Stream
{
   public void WriteHeader(Stream destination)
   {
      ValidateCopyToArguments(destination, bufferSize: 128);

      // Triggers CA1062
      destination.Write(s_Header);
   }
```

I think the CA1062 can be avoided with some hints on the validation call.

Question: I couldn't find the /refs/ for the Stream definition does that exist somewhere and if so does it also need updating?

### CODE REVIEW

**Comment by stephentoub** [comment]
Thanks for your interest in contributing, but I'm not understanding the proposed change.  If you have:
```C#
public void WriteHeader(Stream destination)
```
with nullability checking enabled, that `destination` is already considered non-null.

Can you share a small, complete, compilable repro that demonstrates the issue?

**Comment by CodeBlanch** [comment]
> Can you share a small, complete, compilable repro that demonstrates the issue?

Sure, see: https://github.com/CodeBlanch/stream-nullable-repro

I think the issue is nullability analysis works fine, but CS1062 sees this as a public API so it won't rely on nullability analysis alone (because consumers might not have it enabled)?

**Comment by stephentoub** [comment]
Ah, you meant CA1062 rather than CS1062.

@bartonjs, @terrajobst, what is our recommendation for APIs that validate reference types, and in particular validating that an argument is not null: the proposal here is to make it nullable (even though null isn't allowed) but make it [NotNull] to demonstrate that it's always non-null upon exiting.  This is, in fact, what we do in ArgumentNullException.ThrowIfNull... presumably then that should be our standard and we do the same here for these Validate helpers?

**Comment by stephentoub** [comment]
> I couldn't find the /refs/ for the Stream definition does that exist somewhere and if so does it also need updating?

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/ref/System.Runtime.cs

**Comment by CodeBlanch** [comment]
> Ah, you meant CA1062 rather than CS1062.

Whoops, yes. Sorry about that! Updated the description. Also pushed an update for refs.

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 1]
```diff
@@ -10791,8 +10791,8 @@ public abstract partial class Stream : System.MarshalByRefObject, System.IAsyncD
         public abstract long Seek(long offset, System.IO.SeekOrigin origin);
         public abstract void SetLength(long value);
         public static System.IO.Stream Synchronized(System.IO.Stream stream) { throw null; }
-        protected static void ValidateBufferArguments(byte[] buffer, int offset, int count) { }
-        protected static void ValidateCopyToArguments(System.IO.Stream destination, int bufferSize) { }
+        protected static void ValidateBufferArguments([System.Diagnostics.CodeAnalysis.NotNull] byte[]? buffer, int offset, int count) { throw null; }
+        protected static void ValidateCopyToArguments([System.Diagnostics.CodeAnalysis.NotNull] System.IO.Stream? destination, int bufferSize) { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:10795]:
```suggestion
        protected static void ValidateBufferArguments([System.Diagnostics.CodeAnalysis.NotNullAttribute] byte[]? buffer, int offset, int count) { }
        protected static void ValidateCopyToArguments([System.Diagnostics.CodeAnalysis.NotNullAttribute] System.IO.Stream? destination, int bufferSize) { }
```

**Comment by stephentoub** [comment]
My hesitancy here is that, unlike with ArgumentNullException.ThrowIfNull, we know that this is going to be used with arguments typed as non-nullable; it's then weird that the signature would suggest null inputs are allowed.

Maybe CA1062 should be updated instead to factor in non-nullable parameters and not warn about validating them?
cc: @sharwell

**Comment by CodeBlanch** [comment]
@stephentoub 

>My hesitancy here is that, unlike with ArgumentNullException.ThrowIfNull, we know that this is going to be used with arguments typed as non-nullable; it's then weird that the signature would suggest null inputs are allowed.

I updated them to non-null. Not sure which is the better approach, I can kind of see it both ways. I agree with your hint above that we probably need some agreed-upon convention(s) for validation methods.

**Comment by stephentoub** [comment]
> I updated them to non-null.

They were already not-null, by the fact that they weren't annotated with `?`.  Without the `?`, does this change actually impact anything?  I wouldn't expect it to.

**Comment by terrajobst** [comment]
Looks like we could just mark the parameter as non-null:

```C#
protected void ValidateCopyToArguments([NotNull] Stream destination, int bufferSize);
```

but that feels weird. Equally, mirroring `ThrowIfNull` also seems weird:

```C#
protected void ValidateCopyToArguments([NotNull] Stream? destination, int bufferSize);
```

It would indeed be nicer to fix CA1062.

**Comment by stephentoub** [comment]
> Looks like we could just mark the parameter as non-null:

Does that do anything?  `Stream` is already "non-null".

**Comment by terrajobst** [comment]
> > Looks like we could just mark the parameter as non-null:
> 
> Does that do anything? `Stream` is already "non-null".

Apparently so. When I applied the parameter, CA1062 stopped being raised. Looks like it assumes that if the method returns, the value was non-null, which is also what [we document](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullattribute?view=net-6.0):

> Specifies that an input argument was not null when the call returns.

**Comment by stephentoub** [comment]
> Specifies that an input argument was not null when the call returns.

Yes, but that's intended for input arguments that are maybe null when going in or situations where the parameter is otherwise annotated as maybe null.  It shouldn't be meaningful for parameters that are already not-null, which is the case for a `Stream` rather than `Stream?` parameter.  If that weren't the case, we're missing literally tens of thousands of [NotNull] annotations on practically every non-nullable parameter we have, and obviously we're not going to add that.

This seems like a bug in CA1062. @sharwell? @mavasani?

**Comment by terrajobst** [comment]
> This seems like a bug in CA1062

I agree with that in principle. When a method accepts a non-null parameter, then it is a reasonable assumption that passing it null will result in a crash. Hence, code reachable after wards should assume that it's not null.

**Comment by mavasani** [comment]
CA1062 is not NRT aware, and this is by design https://github.com/dotnet/roslyn-analyzers/issues/3218. CA1062 recognizes 2 attributes for parameters with are validated as non-null: https://github.com/dotnet/roslyn-analyzers/blob/086d1cc888ca0d8c96b94444935f488fcaf3fe98/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.ParameterValidationDataFlowOperationVisitor.cs#L122-L124

**Comment by stephentoub** [comment]
> CA1062 is not NRT aware, and this is by design dotnet/roslyn-analyzers#3218. CA1062 recognizes 2 attributes for parameters with are validated as non-null

NotNullAttribute is part of the NRT system: by recognizing it, CA1062 is choosing to recognize some of NRT but not all of it.  That leads to inconsistencies like the one highlighted here. From my perspective, CA1062 should either respect all of NRT or none of it.

**Comment by mavasani** [comment]
@stephentoub You are correct. Seems original CA1062 implementation did not support NotNull attribute, but this support was added just last year with https://github.com/dotnet/roslyn-analyzers/pull/4249. This leaves us in a partially supported state, which is not good. I think we should take this discussion to runtime design meeting to identify the desired semantics of CA1062 in presence of NRT. Do we have a tracking issue in the runtime repo for this? Otherwise, I can re-open and transfer one of the won't fixed issues in roslyn-analyzers repo.

**Comment by stephentoub** [comment]
> Do we have a tracking issue in the runtime repo for this? Otherwise, I can re-open and transfer one of the won't fixed issues in roslyn-analyzers repo.

We don't have one.  If you could open one or transfer one, that'd be great.
cc: @jeffhandley 

In the meantime, @CodeBlanch, thanks for getting the conversation started.  As we've settled on the right answer being to leave this API the way it is, I'm going to close the PR.

**Pull request closed without merging**